File: resources/gcphone/client/mdt.lua
--====================================================================================
-- #Author: Olle Strand | Cünter#1634
-- #Version 0.1
--====================================================================================

RegisterNetEvent("gcPhone:mdt_login")
AddEventHandler("gcPhone:mdt_login", function(username, password, work, id, adminlevel)
  SendNUIMessage({
    event = 'mdt_login',
      username = username,
      password = password,
      work = work,
      id = id,
      adminlevel = adminlevel
  })
end)

RegisterNetEvent("gcPhone:mdt_updateCitizen")
AddEventHandler("gcPhone:mdt_updateCitizen", function(firstname, lastname, dateofbirth, sex, height, identifier, license)
  SendNUIMessage({
    event = 'mdt_updateCitizen',
      firstname = firstname,
      lastname = lastname,
      dateofbirth = dateofbirth,
      sex = sex,
      height = height,
      identifier = identifier,
      license = license
  })
end)

RegisterNetEvent("gcPhone:mdt_updateVehicle")
AddEventHandler("gcPhone:mdt_updateVehicle", function(firstname, lastname, plate, vehicle)
  local hashVehicule = vehicle.model
	local aheadVehName = GetDisplayNameFromVehicleModel(hashVehicule)
  SendNUIMessage({
    event = 'mdt_updateVehicle',
      firstname = firstname,
      lastname = lastname,
      plate = plate,
      model = aheadVehName
  })
end)

RegisterNetEvent("gcPhone:mdt_addJob")
AddEventHandler("gcPhone:mdt_addJob", function (message, department)
  local pos = GetEntityCoords(PlayerPedId())
  message = message .. " | Sent from GPS Coordinates: " .. pos.x .. ", " .. pos.y

  TriggerServerEvent('gcPhone:addJobToDB', message, department, pos.x, pos.y, pos.z)
end)

RegisterNetEvent("gcPhone:mdt_requestJobs")
AddEventHandler("gcPhone:mdt_requestJobs", function(message, department, isAssigned, coordX, coordY, coordZ, id)
  local assigned = false
  if isAssigned == '1' then
    assigned = true
  end

  SendNUIMessage({
    event = 'mdt_addJob',
      message = message,
      department = department,
      isAssigned = assigned,
      coordX = coordX,
      coordY = coordY,
      coordZ = coordZ,
      jobID = id
  })
end)

allBlips = {}
AddEventHandler("gcPhone:selectJob", function(job, user)
  local x = tonumber(job.coordX)
  local y = tonumber(job.coordY)
  local z = tonumber(job.coordZ)

  Citizen.CreateThread(function()
    local Blip = AddBlipForCoord(x, y, z)
    SetBlipSprite (Blip, 459)
    SetBlipDisplay(Blip, 4)
    SetBlipScale  (Blip, 0.9)
    SetBlipColour (Blip, 6)
    SetBlipAsShortRange(Blip, true)

    BeginTextCommandSetBlipName("STRING")
    AddTextComponentString("MDT Job")
    EndTextCommandSetBlipName(Blip)
    table.insert(allBlips, blip)
  end)
  SetNewWaypoint(x, y)
end)

AddEventHandler("removeBlips", function()
  for i, blip in pairs(allBlips) do
		RemoveBlip(blip)
	end
  allBlips = {}
end)

--====================================================================================
-- #NUI Callbacks
--====================================================================================

RegisterNUICallback('mdt_loginRequest', function(data, cb)
  TriggerServerEvent('gcPhone:mdt_loginRequest', data.username, data.password)
end)

RegisterNUICallback('mdt_citizenRequest', function(data, cb)
  TriggerServerEvent('gcPhone:mdt_citizenRequest', data.firstname, data.lastname)
end)

RegisterNUICallback('mdt_vehicleRequest', function(data, cb)
  TriggerServerEvent('gcPhone:mdt_vehicleRequest', data.plate)
end)

RegisterNUICallback('mdt_requestJobs', function(data, cb)
  TriggerServerEvent('gcPhone:mdt_requestJobs', data.department)
end)

RegisterNUICallback('mdt_jobSelected', function(data, cb)
  TriggerServerEvent('gcPhone:mdt_updateJob', data.job, data.user)
  if allBips ~= {} then
    TriggerEvent('removeBlips')
  end
  TriggerEvent('gcPhone:selectJob', data.job, data.user)
end)

RegisterNUICallback('mdt_jobRemove', function(data, cb)
  TriggerServerEvent('gcPhone:mdt_removeJob', data.job)
end)

File: src_htmlPhone/src/store/modules/mdt.js
import PhoneAPI from './../../PhoneAPI'

const state = {
  mdtAccount: {},
  mdtJobs: [],
  mdtCitizen: {},
  mdtVehicle: {},
  mdtUsers: []
}

const getters = {
  mdtAccount: ({ mdtAccount }) => mdtAccount,
  mdtJobs: ({ mdtJobs }) => mdtJobs,
  mdtCitizen: ({ mdtCitizen }) => mdtCitizen,
  mdtVehicle: ({ mdtVehicle }) => mdtVehicle,
  mdtUsers: ({ mdtUsers }) => mdtUsers
}

const actions = {
  mdtResetData ({ commit }) {
    commit('RESET_DATA', {type: undefined})
  },
  mdtLoginRequest (state, { username, password }) {
    PhoneAPI.mdtLoginRequest(username, password)
  },
  mdtCitizenRequest (state, { firstname, lastname }) {
    PhoneAPI.mdtCitizenRequest(firstname, lastname)
  },
  mdtVehicleRequest (state, { plate }) {
    PhoneAPI.mdtVehicleRequest(plate)
  },
  mdtJobsRequest (state, { department }) {
    PhoneAPI.mdtJobsRequest(department)
  },
  mdtJobSelected (state, { job, user }) {
    PhoneAPI.mdtJobSelected(job, user)
  },
  mdtJobComplete (state, { job }) {
    PhoneAPI.mdtJobRemove(job)
  },
  mdtUpdateAccount (state, { username, password }) {
    PhoneAPI.mdtUpdateAccount(username, password)
  },
  mdtUsersRequest (state, { work, adminlevel }) {
    PhoneAPI.mdtUsersRequest(work, adminlevel)
  },
  mdtResetPassword (state, { user, id }) {
    PhoneAPI.mdtResetPassword(user, id)
  },
  mdtLog (state, { message }) {
    console.log('Log: ' + message)
  },
  mdtLogin ({ commit }, data) {
    var account = {
      username: data.username,
      password: data.password,
      work: data.work,
      id: data.id,
      adminlevel: data.adminlevel
    }
    commit('UPDATE_ACC', account)
  },
  mdtUpdateCitizen ({ commit }, data) {
    var sex_ = ''
    if (data.sex === 'm') {
      sex_ = 'Male'
    } else {
      sex_ = 'Female'
    }
    var citizen = {
      firstname: data.firstname,
      lastname: data.lastname,
      dateofbirth: data.dateofbirth,
      sex: sex_,
      height: data.height,
      identifier: data.identifier,
      license: data.license
    }
    commit('UPDATE_CITIZEN', citizen)
  },
  mdtUpdateVehicle ({ commit }, data) {
    var vehicle = {
      firstname: data.firstname,
      lastname: data.lastname,
      plate: data.plate,
      model: data.model
    }
    commit('UPDATE_VEHICLE', vehicle)
  },
  mdtAddJob ({ commit }, data) {
    var job = {
      message: data.message,
      department: data.department,
      isAssigned: data.isAssigned,
      coordX: data.coordX,
      coordY: data.coordY,
      coordZ: data.coordZ,
      jobID: data.jobID
    }
    commit('MDT_ADD_JOB', job)
  },
  mdtAddUser ({ commit }, data) {
    var user = {
      username: data.username,
      password: data.password,
      work: data.work,
      id: data.id,
      adminlevel: data.adminlevel
    }
    commit('MDT_ADD_USER', user)
  }
}

const mutations = {
  UPDATE_ACC (state, { username, password, work, id, adminlevel }) {
    state.mdtAccount = {
      username: username,
      password: password,
      work: work,
      id: id,
      adminlevel: adminlevel
    }
  },
  UPDATE_CITIZEN (state, { firstname, lastname, dateofbirth, sex, height, identifier, license }) {
    state.mdtCitizen = {
      firstname: firstname,
      lastname: lastname,
      dateofbirth: dateofbirth,
      sex: sex,
      height: height,
      identifier: identifier,
      license: license
    }
  },
  UPDATE_VEHICLE (state, { firstname, lastname, plate, model }) {
    state.mdtVehicle = {
      firstname: firstname,
      lastname: lastname,
      plate: plate,
      model: model
    }
  },
  MDT_ADD_JOB (state, { message, department, isAssigned, coordX, coordY, coordZ, jobID }) {
    state.mdtJobs.unshift({
      message: message,
      department: department,
      isAssigned: isAssigned,
      coordX: coordX,
      coordY: coordY,
      coordZ: coordZ,
      jobID: jobID
    })
  },
  MDT_ADD_USER (state, { username, password, work, id, adminlevel }) {
    state.mdtUsers.unshift({
      username: username,
      password: password,
      work: work,
      id: id,
      adminlevel: adminlevel
    })
  },
  RESET_DATA (state, {type}) {
    state.mdtJobs = []
    state.mdtCitizen = {}
    state.mdtVehicle = {}
    state.mdtUsers = []
  }
}

export default {
  state,
  getters,
  actions,
  mutations
}
